File: content/snippets/extract-product-data-from-a-shop-page/n3.test.py
import json
import time

import pytest

from _harness.fake_llm import FakeLLM, FakeLLMError
from _harness.fake_sdk import FakeSDK
from n3 import (FIELDS, MAX_CHARACTERS, MAX_HTML, MODEL, ProviderClient, ReadingUnavailable,
                read_product, to_text)

PAGE = (
    "<!doctype html><html><head><style>.prix{color:red}</style>"
    "<script>window.tracking={}</script></head><body>"
    "<nav>Accueil  Boutique</nav>"
    "<h1>Moulin à café Lumière</h1><p>Référence MC-4501</p>"
    '<p class="prix">19,90 €</p><p>En stock</p></body></html>'
)

LU = {"name": "Moulin à café Lumière", "sku": "MC-4501", "brand": "Lumière",
      "price": "19,90", "currency": "EUR", "availability": "En stock"}


def double(reponse=None, **extra) -> FakeLLM:
    return FakeLLM(response=json.dumps(reponse if reponse is not None else LU,
                                       ensure_ascii=False), **extra)


# ---------------------------------------------------------------------------
# Point de rupture
# ---------------------------------------------------------------------------


def test_point_de_rupture_la_garde_verifie_dou_vient_une_valeur_pas_ce_quelle_veut_dire():
    """
    « La devise écrite « € » dans la page et rendue « EUR » par le modèle est
    comptée comme inventée, alors qu'elle est juste. »
    """
    rapport = read_product(PAGE, double())
    assert "€" in PAGE and "EUR" not in PAGE
    assert rapport["invented"] == ["currency"]
    assert rapport["product"]["currency"] is None


def test_point_de_rupture_temoin_un_prix_absent_de_la_page_est_bien_ecarte():
    """
    « Le témoin : un prix que la page ne contient pas — « 24,90 » sur une page
    à 19,90 — est bien écarté. »
    """
    rapport = read_product(PAGE, double({**LU, "price": "24,90"}))
    assert "24,90" not in PAGE
    assert rapport["product"]["price"] is None
    assert "price" in rapport["invented"]
    # Témoin du témoin : le prix de la page, lui, est gardé.
    assert read_product(PAGE, double())["product"]["price"] == "19,90"


# ---------------------------------------------------------------------------
# Les autres affirmations du niveau
# ---------------------------------------------------------------------------


def test_la_page_est_debarrassee_de_son_balisage_avant_de_partir():
    """Docstring : « the tags come off and what is left is capped »."""
    texte = to_text(PAGE)
    assert "<" not in texte and ">" not in texte
    assert "window.tracking" not in texte and "color:red" not in texte
    assert "Moulin à café Lumière" in texte and "19,90 €" in texte
    assert len(texte) < len(PAGE) / 3


def test_la_page_est_coupee_au_budget_pas_refusee():
    """
    Docstring : « A page that does not fit is cut, not refused: a scraper that
    raises on a long page returns nothing at all ».
    """
    enorme = PAGE.replace("</body>", "<p>" + "x" * 500_000 + "</p></body>")
    rapport = read_product(enorme, double())
    assert rapport["characters_sent"] == MAX_CHARACTERS
    assert rapport["product"]["name"] == "Moulin à café Lumière"


def test_la_requete_porte_le_texte_de_la_page_et_les_champs_demandes():
    client = double()
    read_product(PAGE, client)
    envoye = client.last_request["prompt"]
    assert "Moulin à café Lumière" in envoye
    for champ in FIELDS:
        assert champ in envoye
    assert client.last_request["temperature"] == 0


def test_une_reponse_dans_une_cloture_de_code_est_decodee():
    """Charte des tests : une seule clôture de code se décode comme si elle n'y était pas."""
    client = FakeLLM(response="```json\n" + json.dumps(LU, ensure_ascii=False) + "\n```")
    assert read_product(PAGE, client)["product"]["sku"] == "MC-4501"
    # Deux blocs, ou du texte autour, ne se décodent pas.
    for mauvaise in ["Voici : " + json.dumps(LU), "```json\n{}\n```\n```json\n{}\n```",
                     "```json\n{}", ""]:
        with pytest.raises(ReadingUnavailable):
            read_product(PAGE, FakeLLM(response=mauvaise), attempts=1)


def test_un_refus_du_modele_nest_pas_passe_au_decodeur():
    """Docstring : « `content` of None is not a value »."""
    with pytest.raises(ReadingUnavailable) as leve:
        read_product(PAGE, FakeLLM(response=None), attempts=1)
    assert "the model answered no text" in str(leve.value)


def test_une_panne_est_retentee_le_nombre_de_fois_annonce():
    client = double(fail_times=2)
    assert read_product(PAGE, client, attempts=3)["product"]["sku"] == "MC-4501"
    assert client.call_count == 3
    # Une panne de plus que d'essais, et l'erreur remonte.
    trop = double(fail_times=3)
    with pytest.raises(ReadingUnavailable):
        read_product(PAGE, trop, attempts=3)
    assert trop.call_count == 3


def test_ladaptateur_par_defaut_parle_au_vrai_kit():
    """T2 : l'adaptateur est exécuté contre le double du harnais."""
    sdk = FakeSDK(content=json.dumps(LU, ensure_ascii=False))
    client = ProviderClient(sdk=sdk)
    rapport = read_product(PAGE, client)
    assert rapport["product"]["sku"] == "MC-4501"
    envoye = sdk.last_request
    assert envoye["endpoint"] == "chat.completions"
    assert envoye["model"] == MODEL
    assert envoye["temperature"] == 0
    assert envoye["messages"][0]["role"] == "user"
    assert "Moulin à café Lumière" in envoye["messages"][0]["content"]
    # Un refus du fournisseur : content nul, lu comme tel.
    with pytest.raises(ReadingUnavailable):
        read_product(PAGE, ProviderClient(sdk=FakeSDK(content=None)), attempts=1)


# ---------------------------------------------------------------------------
# Cas de production
# ---------------------------------------------------------------------------


def test_production_entree_banale_une_fiche_sans_donnees_structurees():
    """T5 : l'entrée ordinaire du public visé — la boutique qui n'émet rien."""
    rapport = read_product(PAGE, double())
    assert rapport["source"] == "model"
    assert rapport["product"]["name"] == "Moulin à café Lumière"
    assert rapport["product"]["sku"] == "MC-4501"


def test_production_entree_vide():
    rapport = read_product("", double({f: None for f in FIELDS}))
    assert rapport["characters_sent"] == 0
    assert all(v is None for v in rapport["product"].values())
    assert rapport["invented"] == []


def test_production_encodages_inattendus():
    # Emoji, insécables, entités : le texte envoyé les garde tels quels.
    page = PAGE.replace("Moulin à café Lumière", "Bouilloire 🫖 Lumière")
    rapport = read_product(page, double({**LU, "name": "Bouilloire 🫖 Lumière"}))
    assert rapport["product"]["name"] == "Bouilloire 🫖 Lumière"
    # La comparaison ignore la casse, pas les accents.
    assert read_product(PAGE, double({**LU, "name": "MOULIN À CAFÉ LUMIÈRE"}))[
        "product"]["name"] == "MOULIN À CAFÉ LUMIÈRE"
    assert "name" in read_product(PAGE, double({**LU, "name": "Moulin a cafe Lumiere"}))["invented"]


def test_production_entree_tres_grande_la_page_hostile_du_niveau_N0():
    """
    La page que le niveau N0 de cette fiche est écrit pour survivre : vingt
    mille balises `<script>` jamais refermées. Le nettoyage de ce niveau la
    traversait en quarante secondes de processeur, **avant** l'appel au
    modèle — c'est-à-dire qu'il réintroduisait, deux fichiers plus loin,
    l'effondrement que le N0 évite.

    La charte des tests demande qu'une entrée trop grande soit refusée avant
    l'appel. Le plafond de balisage est donc appliqué en premier, et les blocs
    sont balayés plutôt que reconnus par une expression régulière.
    """
    hostile = '<script type="application/ld+json">' * 20_000
    debut = time.perf_counter()
    assert to_text(hostile) == ""
    assert time.perf_counter() - debut < 1.0
    # Les mêmes balises, en `<style>` : c'est la seconde branche du motif.
    hostile_style = "<style>" * 20_000
    debut = time.perf_counter()
    assert to_text(hostile_style) == ""
    assert time.perf_counter() - debut < 1.0
    # Témoin : une page ordinaire, cent fois le cas nominal, reste lisible.
    grande = PAGE + "<p>Livraison offerte.</p>" * 10_000
    debut = time.perf_counter()
    texte = to_text(grande)
    assert time.perf_counter() - debut < 1.0
    assert "Moulin à café Lumière" in texte


def test_production_le_plafond_de_balisage_sapplique_avant_le_nettoyage():
    """
    Commentaire : « a cap applied after the cleaning protects nothing, since
    the cleaning is the part that costs ».
    """
    # Au-delà du plafond, le balisage n'est pas lu du tout : le nom qui suit
    # ne peut pas ressortir.
    loin = "<p>a</p>" * (MAX_HTML // 8 + 10) + "<p>Moulin à café Lumière</p>"
    assert len(loin) > MAX_HTML
    assert "Moulin" not in to_text(loin)
    # Juste en dessous, il l'est.
    proche = "<p>a</p>" * 10 + "<p>Moulin à café Lumière</p>"
    assert len(proche) < MAX_HTML
    assert "Moulin à café Lumière" in to_text(proche)


def test_production_valeurs_aux_limites():
    # Exactement le budget, et un caractère de plus.
    texte = "a" * MAX_CHARACTERS
    assert read_product(texte, double({f: None for f in FIELDS}))["characters_sent"] == MAX_CHARACTERS
    assert read_product(texte + "b", double({f: None for f in FIELDS}))[
        "characters_sent"] == MAX_CHARACTERS
    # Une réponse qui n'est pas un objet, et une réponse sans les clés.
    assert read_product(PAGE, FakeLLM(response="[]"))["product"] == {f: None for f in FIELDS}
    assert read_product(PAGE, FakeLLM(response="{}"))["invented"] == []


def test_production_un_champ_invente_nempeche_pas_de_garder_les_autres():
    """T8 : une valeur fausse ne fait pas tomber la lecture."""
    rapport = read_product(PAGE, double({**LU, "sku": "MC-0000", "price": "24,90"}))
    assert rapport["product"]["name"] == "Moulin à café Lumière"
    assert rapport["invented"] == ["sku", "price", "currency"]


def test_production_la_lecture_tient_la_classe_de_latence_annoncee():
    """
    latency « ~1 s » : la classe est celle de l'appel au fournisseur, que le
    double ne mesure pas. Ce test ne borne que la part locale — découpe du
    balisage et vérification des valeurs — sur mille lectures.
    """
    client = double()
    debut = time.perf_counter()
    for _ in range(1_000):
        read_product(PAGE, client)
    assert time.perf_counter() - debut < 60.0

File: content/snippets/extract-product-data-from-a-shop-page/n3.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import { FakeLLM } from '../_harness/fake-llm.mjs';
import { FakeSDK } from '../_harness/fake-sdk.mjs';
import {
  FIELDS, MAX_CHARACTERS, MAX_HTML, MODEL, ReadingUnavailable, providerClient, readProduct, toText,
} from './n3.js';

const PAGE = '<!doctype html><html><head><style>.prix{color:red}</style>'
  + '<script>window.tracking={}</script></head><body>'
  + '<nav>Accueil  Boutique</nav>'
  + '<h1>Moulin à café Lumière</h1><p>Référence MC-4501</p>'
  + '<p class="prix">19,90 €</p><p>En stock</p></body></html>';

const LU = {
  name: 'Moulin à café Lumière', sku: 'MC-4501', brand: 'Lumière',
  price: '19,90', currency: 'EUR', availability: 'En stock',
};

const double = (reponse = LU, extra = {}) => new FakeLLM({
  response: JSON.stringify(reponse), ...extra,
});

const vide = () => Object.fromEntries(FIELDS.map((f) => [f, null]));

// ---------------------------------------------------------------------------
// Point de rupture
// ---------------------------------------------------------------------------

test("point de rupture : la garde vérifie d'où vient une valeur, pas ce qu'elle veut dire", async () => {
  const rapport = await readProduct(PAGE, { client: double() });
  assert.ok(PAGE.includes('€'));
  assert.ok(!PAGE.includes('EUR'));
  assert.deepEqual(rapport.invented, ['currency']);
  assert.equal(rapport.product.currency, null);
});

test("point de rupture : témoin, un prix absent de la page est bien écarté", async () => {
  const rapport = await readProduct(PAGE, { client: double({ ...LU, price: '24,90' }) });
  assert.ok(!PAGE.includes('24,90'));
  assert.equal(rapport.product.price, null);
  assert.ok(rapport.invented.includes('price'));
  assert.equal((await readProduct(PAGE, { client: double() })).product.price, '19,90');
});

// ---------------------------------------------------------------------------
// Les autres affirmations du niveau
// ---------------------------------------------------------------------------

test('la page est débarrassée de son balisage avant de partir', () => {
  const texte = toText(PAGE);
  assert.ok(!texte.includes('<'));
  assert.ok(!texte.includes('>'));
  assert.ok(!texte.includes('window.tracking'));
  assert.ok(!texte.includes('color:red'));
  assert.ok(texte.includes('Moulin à café Lumière'));
  assert.ok(texte.includes('19,90 €'));
  assert.ok(texte.length < PAGE.length / 3);
});

test('la page est coupée au budget, pas refusée', async () => {
  const enorme = PAGE.replace('</body>', `<p>${'x'.repeat(500_000)}</p></body>`);
  const rapport = await readProduct(enorme, { client: double() });
  assert.equal(rapport.characters_sent, MAX_CHARACTERS);
  assert.equal(rapport.product.name, 'Moulin à café Lumière');
});

test('la requête porte le texte de la page et les champs demandés', async () => {
  const client = double();
  await readProduct(PAGE, { client });
  const envoye = client.lastRequest.prompt;
  assert.ok(envoye.includes('Moulin à café Lumière'));
  for (const champ of FIELDS) assert.ok(envoye.includes(champ), champ);
  assert.equal(client.lastRequest.temperature, 0);
});

test('une réponse dans une clôture de code est décodée', async () => {
  const client = new FakeLLM({ response: `\`\`\`json\n${JSON.stringify(LU)}\n\`\`\`` });
  assert.equal((await readProduct(PAGE, { client })).product.sku, 'MC-4501');
  for (const mauvaise of [`Voici : ${JSON.stringify(LU)}`,
    '```json\n{}\n```\n```json\n{}\n```', '```json\n{}', '']) {
    await assert.rejects(
      () => readProduct(PAGE, { client: new FakeLLM({ response: mauvaise }), attempts: 1 }),
      ReadingUnavailable,
    );
  }
});

test("un refus du modèle n'est pas passé au décodeur", async () => {
  // Un refus revient sans contenu. Le double du harnais rend `null` sous forme
  // de texte, alors ce cas se pose directement : un client dont `complete`
  // rend null, comme le fait l'adaptateur sur un `content` nul.
  const refus = { complete: async () => null };
  await assert.rejects(
    () => readProduct(PAGE, { client: refus, attempts: 1 }),
    (erreur) => erreur instanceof ReadingUnavailable
      && String(erreur.message).includes('the model answered no text'),
  );
});

test("une panne est retentée le nombre de fois annoncé", async () => {
  const client = double(LU, { failTimes: 2 });
  assert.equal((await readProduct(PAGE, { client, attempts: 3 })).product.sku, 'MC-4501');
  assert.equal(client.callCount, 3);
  const trop = double(LU, { failTimes: 3 });
  await assert.rejects(() => readProduct(PAGE, { client: trop, attempts: 3 }), ReadingUnavailable);
  assert.equal(trop.callCount, 3);
});

test("l'adaptateur par défaut parle au vrai kit", async () => {
  // T2 : l'adaptateur est exécuté contre le double du harnais.
  const sdk = new FakeSDK({ content: JSON.stringify(LU) });
  const client = await providerClient(sdk);
  const rapport = await readProduct(PAGE, { client });
  assert.equal(rapport.product.sku, 'MC-4501');
  const envoye = sdk.lastRequest;
  assert.equal(envoye.endpoint, 'chat.completions');
  assert.equal(envoye.model, MODEL);
  assert.equal(envoye.temperature, 0);
  assert.equal(envoye.messages[0].role, 'user');
  assert.ok(envoye.messages[0].content.includes('Moulin à café Lumière'));
  const refus = await providerClient(new FakeSDK({ content: null }));
  await assert.rejects(() => readProduct(PAGE, { client: refus, attempts: 1 }), ReadingUnavailable);
});

// ---------------------------------------------------------------------------
// Cas de production
// ---------------------------------------------------------------------------

test('production : entrée banale, une fiche sans données structurées', async () => {
  const rapport = await readProduct(PAGE, { client: double() });
  assert.equal(rapport.source, 'model');
  assert.equal(rapport.product.name, 'Moulin à café Lumière');
  assert.equal(rapport.product.sku, 'MC-4501');
});

test('production : entrée vide', async () => {
  const rapport = await readProduct('', { client: double(vide()) });
  assert.equal(rapport.characters_sent, 0);
  assert.ok(Object.values(rapport.product).every((v) => v === null));
  assert.deepEqual(rapport.invented, []);
});

test('production : encodages inattendus', async () => {
  const page = PAGE.replace('Moulin à café Lumière', 'Bouilloire 🫖 Lumière');
  const rapport = await readProduct(page, {
    client: double({ ...LU, name: 'Bouilloire 🫖 Lumière' }),
  });
  assert.equal(rapport.product.name, 'Bouilloire 🫖 Lumière');
  const majuscules = await readProduct(PAGE, {
    client: double({ ...LU, name: 'MOULIN À CAFÉ LUMIÈRE' }),
  });
  assert.equal(majuscules.product.name, 'MOULIN À CAFÉ LUMIÈRE');
  const sansAccent = await readProduct(PAGE, {
    client: double({ ...LU, name: 'Moulin a cafe Lumiere' }),
  });
  assert.ok(sansAccent.invented.includes('name'));
});

test('production : entrée très grande, la page hostile du niveau N0', () => {
  // La page que le niveau N0 de cette fiche est écrit pour survivre : vingt
  // mille balises `<script>` jamais refermées. Le nettoyage de ce niveau la
  // traversait avec le même effondrement d'expression régulière que le N0
  // évite, **avant** l'appel au modèle. La charte des tests demande qu'une
  // entrée trop grande soit refusée avant l'appel : le plafond de balisage est
  // donc appliqué en premier, et les blocs sont balayés.
  const hostile = '<script type="application/ld+json">'.repeat(20_000);
  let debut = performance.now();
  assert.equal(toText(hostile), '');
  assert.ok(performance.now() - debut < 1_000);
  // Les mêmes balises, en `<style>` : c'est la seconde branche du motif.
  debut = performance.now();
  assert.equal(toText('<style>'.repeat(20_000)), '');
  assert.ok(performance.now() - debut < 1_000);
  // Témoin : une page ordinaire, cent fois le cas nominal, reste lisible.
  const grande = PAGE + '<p>Livraison offerte.</p>'.repeat(10_000);
  debut = performance.now();
  const texte = toText(grande);
  assert.ok(performance.now() - debut < 1_000);
  assert.ok(texte.includes('Moulin à café Lumière'));
});

test("production : le plafond de balisage s'applique avant le nettoyage", () => {
  // Commentaire : « a cap applied after the cleaning protects nothing, since
  // the cleaning is the part that costs ».
  const loin = '<p>a</p>'.repeat(Math.floor(MAX_HTML / 8) + 10) + '<p>Moulin à café Lumière</p>';
  assert.ok(loin.length > MAX_HTML);
  assert.ok(!toText(loin).includes('Moulin'));
  const proche = '<p>a</p>'.repeat(10) + '<p>Moulin à café Lumière</p>';
  assert.ok(proche.length < MAX_HTML);
  assert.ok(toText(proche).includes('Moulin à café Lumière'));
});

test('production : valeurs aux limites', async () => {
  const texte = 'a'.repeat(MAX_CHARACTERS);
  assert.equal(
    (await readProduct(texte, { client: double(vide()) })).characters_sent, MAX_CHARACTERS,
  );
  assert.equal(
    (await readProduct(`${texte}b`, { client: double(vide()) })).characters_sent, MAX_CHARACTERS,
  );
  assert.deepEqual(
    (await readProduct(PAGE, { client: new FakeLLM({ response: '[]' }) })).product, vide(),
  );
  assert.deepEqual(
    (await readProduct(PAGE, { client: new FakeLLM({ response: '{}' }) })).invented, [],
  );
});

test("production : un champ inventé n'empêche pas de garder les autres", async () => {
  const rapport = await readProduct(PAGE, {
    client: double({ ...LU, sku: 'MC-0000', price: '24,90' }),
  });
  assert.equal(rapport.product.name, 'Moulin à café Lumière');
  assert.deepEqual(rapport.invented, ['sku', 'price', 'currency']);
});

test('production : la lecture tient la classe de latence annoncée', async () => {
  // latency « ~1 s » : la classe est celle de l'appel au fournisseur, que le
  // double ne mesure pas. Ce test ne borne que la part locale.
  const client = double();
  const debut = performance.now();
  for (let i = 0; i < 1_000; i += 1) await readProduct(PAGE, { client });
  assert.ok(performance.now() - debut < 60_000);
});
